src/book/02-system/02-subsystems/06-logging.mdx

```diff
@@ -0,0 +1,42 @@
+---
+section: System
+chapter: Subsystems
+title: Logging
+description: Logging and how it works in the NUbots codebase.
+slug: /system/subsystems/logging
+---
+
+# Data Playback
+
+## Description
+
+DataPlayback is a module which will allow you to play back `.nbs` files as if they were happening in realtime. This can be very useful for when you are trying to diagnose a problem for which you have a recording as you can continuously play back the same data.
+
+The module will buffer up packets from the future and emit them at the same rate that they were emitted originally.
+
+## Usage
+
+Add `support::logging::DataPlayback` to your role file and comment out the roles that you want the `nbs` files to replace. For example, if you are playing back `message.input.Sensors` then you need to comment out `platform::darwin::SensorFilter` from your role otherwise you will have two different modules emitting conflicting `Sensors` messages. You will now need to recompile your role for these changes to take effect.
```

Comment:
```suggestion
Add `support::logging::DataPlayback` to your role file and comment out the roles that you want the `nbs` files to replace. For example, if you are playing back `message.input.Sensors` you need to comment out `platform::darwin::SensorFilter` from your role otherwise you will have two different modules emitting conflicting `Sensors` messages. You must recompile your role for these changes to take effect.
```